File: client/src/services/treatmentHistoryService.ts
import api from './api';
import type { ApiResponse } from '../types/api';

export interface TreatmentHistory {
  id: string;
  clinicId: string;
  patientId: string;
  patientName: string;
  appointmentId?: string;
  staffId: string;
  staffName: string;
  treatmentDate: string;
  chiefComplaint?: string;
  symptoms?: string;
  bloodPressure?: string;
  temperature?: number;
  heartRate?: number;
  respiratoryRate?: number;
  weight?: number;
  height?: number;
  physicalExamination?: string;
  diagnosis?: string;
  differentialDiagnosis?: string;
  treatment: string;
  prescriptionNotes?: string;
  treatmentPlan?: string;
  followUpInstructions?: string;
  nextAppointmentDate?: string;
  notes?: string;
  createdAt: string;
}

export interface CreateTreatmentHistoryRequest {
  patientId: string;
  appointmentId?: string;
  treatmentDate: string;
  chiefComplaint?: string;
  symptoms?: string;
  bloodPressure?: string;
  temperature?: number;
  heartRate?: number;
  respiratoryRate?: number;
  weight?: number;
  height?: number;
  physicalExamination?: string;
  diagnosis?: string;
  differentialDiagnosis?: string;
  treatment: string;
  prescriptionNotes?: string;
  treatmentPlan?: string;
  followUpInstructions?: string;
  nextAppointmentDate?: string;
  notes?: string;
}

export interface UpdateTreatmentHistoryRequest {
  chiefComplaint?: string;
  symptoms?: string;
  bloodPressure?: string;
  temperature?: number;
  heartRate?: number;
  respiratoryRate?: number;
  weight?: number;
  height?: number;
  physicalExamination?: string;
  diagnosis?: string;
  differentialDiagnosis?: string;
  treatment: string;
  prescriptionNotes?: string;
  treatmentPlan?: string;
  followUpInstructions?: string;
  nextAppointmentDate?: string;
  notes?: string;
}

export interface TreatmentHistoryListResponse {
  items: TreatmentHistory[];
  total: number;
  page: number;
  pageSize: number;
}

export const treatmentHistoryService = {
  getTreatmentHistory: async (patientId?: string, search?: string, page = 1, pageSize = 10): Promise<ApiResponse<TreatmentHistoryListResponse>> => {
    const params = new URLSearchParams();
    if (patientId) params.append('patientId', patientId);
    if (search) params.append('search', search);
    params.append('page', page.toString());
    params.append('pageSize', pageSize.toString());
    
    const response = await api.get<ApiResponse<TreatmentHistoryListResponse>>(`/treatmenthistory?${params}`);
    return response.data;
  },

  getPatientTreatmentHistory: async (patientId: string): Promise<ApiResponse<TreatmentHistory[]>> => {
    const response = await api.get<ApiResponse<TreatmentHistory[]>>(`/treatmenthistory/patient/${patientId}`);
    return response.data;
  },

  getTreatmentByAppointment: async (appointmentId: string): Promise<ApiResponse<TreatmentHistory>> => {
    const response = await api.get<ApiResponse<TreatmentHistory>>(`/treatmenthistory/appointment/${appointmentId}`);
    return response.data;
  },

  getTreatment: async (id: string): Promise<ApiResponse<TreatmentHistory>> => {
    const response = await api.get<ApiResponse<TreatmentHistory>>(`/treatmenthistory/${id}`);
    return response.data;
  },

  createTreatment: async (data: CreateTreatmentHistoryRequest): Promise<ApiResponse<TreatmentHistory>> => {
    const response = await api.post<ApiResponse<TreatmentHistory>>('/treatmenthistory', data);
    return response.data;
  },

  updateTreatment: async (id: string, data: UpdateTreatmentHistoryRequest): Promise<ApiResponse<TreatmentHistory>> => {
    const response = await api.put<ApiResponse<TreatmentHistory>>(`/treatmenthistory/${id}`, data);
    return response.data;
  },

  deleteTreatment: async (id: string): Promise<ApiResponse<void>> => {
    const response = await api.delete<ApiResponse<void>>(`/treatmenthistory/${id}`);
    return response.data;
  },
};

File: src/ClinicManagement.Infrastructure/Services/TreatmentHistoryService.cs
using ClinicManagement.Core.Entities;
using ClinicManagement.Core.Interfaces;

namespace ClinicManagement.Infrastructure.Services;

public class TreatmentHistoryService : ITreatmentHistoryService
{
    private readonly ITreatmentHistoryRepository _treatmentHistoryRepository;
    private readonly IClinicContext _clinicContext;

    public TreatmentHistoryService(ITreatmentHistoryRepository treatmentHistoryRepository, IClinicContext clinicContext)
    {
        _treatmentHistoryRepository = treatmentHistoryRepository;
        _clinicContext = clinicContext;
    }

    public async Task<(bool Success, string? ErrorCode, TreatmentHistory? TreatmentHistory)> CreateTreatmentAsync(
        Guid patientId,
        Guid? appointmentId,
        Guid staffId,
        DateTime treatmentDate,
        string? chiefComplaint,
        string? symptoms,
        string? bloodPressure,
        decimal? temperature,
        int? heartRate,
        int? respiratoryRate,
        decimal? weight,
        decimal? height,
        string? physicalExamination,
        string? diagnosis,
        string? differentialDiagnosis,
        string treatment,
        string? prescription,
        string? treatmentPlan,
        string? followUpInstructions,
        DateTime? nextAppointmentDate,
        string? notes)
    {
        if (!_clinicContext.CurrentClinicId.HasValue)
        {
            return (false, "AUTH_UNAUTHORIZED", null);
        }

        var treatmentHistory = new TreatmentHistory
        {
            ClinicId = _clinicContext.CurrentClinicId.Value,
            PatientId = patientId,
            AppointmentId = appointmentId,
            StaffId = staffId,
            TreatmentDate = treatmentDate,
            ChiefComplaint = chiefComplaint,
            Symptoms = symptoms,
            BloodPressure = bloodPressure,
            Temperature = temperature,
            HeartRate = heartRate,
            RespiratoryRate = respiratoryRate,
            Weight = weight,
            Height = height,
            PhysicalExamination = physicalExamination,
            Diagnosis = diagnosis,
            DifferentialDiagnosis = differentialDiagnosis,
            Treatment = treatment,
            PrescriptionNotes = prescription,
            TreatmentPlan = treatmentPlan,
            FollowUpInstructions = followUpInstructions,
            NextAppointmentDate = nextAppointmentDate,
            Notes = notes
        };

        await _treatmentHistoryRepository.AddAsync(treatmentHistory);
        await _treatmentHistoryRepository.SaveChangesAsync();

        return (true, null, treatmentHistory);
    }

    public async Task<(bool Success, string? ErrorCode, TreatmentHistory? TreatmentHistory)> UpdateTreatmentAsync(
        Guid id,
        string? chiefComplaint,
        string? symptoms,
        string? bloodPressure,
        decimal? temperature,
        int? heartRate,
        int? respiratoryRate,
        decimal? weight,
        decimal? height,
        string? physicalExamination,
        string? diagnosis,
        string? differentialDiagnosis,
        string treatment,
        string? prescription,
        string? treatmentPlan,
        string? followUpInstructions,
        DateTime? nextAppointmentDate,
        string? notes)
    {
        var treatmentHistory = await _treatmentHistoryRepository.GetByIdAsync(id);
        if (treatmentHistory == null)
        {
            return (false, "NOT_FOUND", null);
        }

        treatmentHistory.ChiefComplaint = chiefComplaint;
        treatmentHistory.Symptoms = symptoms;
        treatmentHistory.BloodPressure = bloodPressure;
        treatmentHistory.Temperature = temperature;
        treatmentHistory.HeartRate = heartRate;
        treatmentHistory.RespiratoryRate = respiratoryRate;
        treatmentHistory.Weight = weight;
        treatmentHistory.Height = height;
        treatmentHistory.PhysicalExamination = physicalExamination;
        treatmentHistory.Diagnosis = diagnosis;
        treatmentHistory.DifferentialDiagnosis = differentialDiagnosis;
        treatmentHistory.Treatment = treatment;
        treatmentHistory.PrescriptionNotes = prescription;
        treatmentHistory.TreatmentPlan = treatmentPlan;
        treatmentHistory.FollowUpInstructions = followUpInstructions;
        treatmentHistory.NextAppointmentDate = nextAppointmentDate;
        treatmentHistory.Notes = notes;

        await _treatmentHistoryRepository.UpdateAsync(treatmentHistory);
        await _treatmentHistoryRepository.SaveChangesAsync();

        return (true, null, treatmentHistory);
    }

    public async Task<(bool Success, string? ErrorCode)> DeleteTreatmentAsync(Guid id)
    {
        var treatmentHistory = await _treatmentHistoryRepository.GetByIdAsync(id);
        if (treatmentHistory == null)
        {
            return (false, "NOT_FOUND");
        }

        await _treatmentHistoryRepository.SoftDeleteAsync(treatmentHistory);
        await _treatmentHistoryRepository.SaveChangesAsync();

        return (true, null);
    }

    public async Task<TreatmentHistory?> GetTreatmentByIdAsync(Guid id)
    {
        return await _treatmentHistoryRepository.GetByIdAsync(id);
    }

    public async Task<TreatmentHistory?> GetTreatmentByAppointmentIdAsync(Guid appointmentId)
    {
        return await _treatmentHistoryRepository.GetByAppointmentIdAsync(appointmentId);
    }

    public async Task<IEnumerable<TreatmentHistory>> GetPatientTreatmentHistoryAsync(Guid patientId)
    {
        return await _treatmentHistoryRepository.GetPatientTreatmentHistoryAsync(patientId);
    }

    public async Task<(IEnumerable<TreatmentHistory> Items, int Total)> SearchTreatmentHistoryAsync(Guid? patientId, string? search, int page, int pageSize)
    {
        var items = await _treatmentHistoryRepository.SearchTreatmentHistoryAsync(patientId, search, page, pageSize);
        var total = await _treatmentHistoryRepository.GetTotalCountAsync(patientId, search);
        return (items, total);
    }
}